guix: Misc feedback-based fixes + hier restructuring

dongcarl opened the pull request:
This PR addresses a few hiccups encountered by the brave souls who've been experimenting with the Guix scripts:
- Resolves confusion between `--cores=` and `--max-jobs=`
  - `guix`'s `--cores=` actually corresponds to make's `--jobs=`, so let's just control `--cores=` with our overridable env var
- `git-describe` will scream `fatal: no tag exactly matches '<hash>'` when looking for a tag, but we don't care, so silence that
- `windeploy/unsigned` should be inside `distsrc-*` and created idempotently (sorry I know this one annoyed people)
- Add troubleshooting documentation to `README.md`
- Add early health check for `guix-daemon` in case user forgot to start a `guix-daemon`
- Depending on configuration, a `--fallback` flag may be needed to tell Guix to not fail if substitutes fail but fallback to building locally
- `codesign_allocate` and `pagestuff` are now unnecessary for codesigning as we're now using `signapple`

A few robustness changes are also included:
- We supply the `--link-profile` flag, as some Guix packages may expect the profile to be available under `$HOME/.guix-profile`
- We now clear and manually set all toolchain-related env vars (e.g. `C*_INCLUDE_PATH`) ourselves, after patching a Qt::moc bug
- We use the native `clang-toolchain` package for darwin builds instead of `clang`, lining up with all our other toolchain packages.

Finally, we restructure the guix building hierarchy such that it looks something like:
```
guix-build-<short-hash-or-version-tag>
├── distsrc-<short-hash-or-version-tag>-${HOST}
│   ├── contrib
│   ├── depends
│   ├── src
│   └── ...
├── distsrc-<short-hash-or-version-tag>-...
└── output
    ├── dist-archive
    │   └── bitcoin-<short-hash-or-version-tag>.tar.gz
    ├── *-linux-*
    │   ├── bitcoin-<short-hash-or-version-tag>-*-linux-*-debug.tar.gz
    │   └── bitcoin-<short-hash-or-version-tag>-*-linux-*.tar.gz
    ├── x86_64-apple-darwin18
    │   ├── bitcoin-<short-hash-or-version-tag>-osx64.tar.gz
    │   ├── bitcoin-<short-hash-or-version-tag>-osx-unsigned.dmg
    │   └── bitcoin-<short-hash-or-version-tag>-osx-unsigned.tar.gz
    └── x86_64-w64-mingw32
        ├── bitcoin-<short-hash-or-version-tag>-win64-debug.zip
        ├── bitcoin-<short-hash-or-version-tag>-win64-setup-unsigned.exe
        ├── bitcoin-<short-hash-or-version-tag>-win64.zip
        └── bitcoin-<short-hash-or-version-tag>-win-unsigned.tar.gz
```
Separating guix builds by their version identifier (basically namespacing them) allows us to change the layout in the future without worry about potential naming conflicts.

dongcarl committed:
guix: Use --cores instead of --max-jobs

In Guix, there are two flags for controlling parallelism:

Note: When I say "derivation," think "package"

--cores=n
  - controls the number of CPU cores to build each derivation. This is
    the value passed to `make`'s `--jobs=` flag.
  - defaults to 0: as many cores as is available

--max-jobs=n
  - controls how many derivations can be built in parallel
  - defaults to 1

Therefore, if set --max-jobs=$MAX_JOBS and don't set --cores, Guix could
theoretically spin up $MAX_JOBS * $(nproc) number of threads, and that's
no good.

So we could either default to --cores=1, --max-jobs=$MAX_JOBS

  - Pro: --cores=1 means that `make` will be invoked with `-j1`,
         avoiding problems with package whose build systems and test
         suites break when running multi-threaded.

  - Con: There will be times when only 1 or 2 derivations can be built
         at a time, because the rest of the dependency graph all depend
         on those 1 or 2 derivations. During these times, the machine
         will be severely under-utilized.

or --cores=$MAX_JOBS, --max-jobs=1

  - Pro: We don't encounter prolonged periods of
         severe under-utilization mentioned above.

  - Con: Many packages' build systems and test suites break when running
         multi-threaded.

or --cores=1, --max-jobs=1 and let the user override with
$ADDITIONAL_GUIX_COMMON_FLAGS

dongcarl committed:
contrib: Silence git-describe when looking for tag

Otherwise, it prints a rather disturbing message to stderr:

    fatal: no tag exactly matches '<hash>'

dongcarl committed:
guix: Create windeploy inside distsrc-*

./windeploy is a "working directory", and therefore belongs inside
distsrc-*. Many people have noticed their Guix builds failing after
hours simply because they did not remove windeploy (but did remove the
distsrc-* directories).

dongcarl committed:
guix: Add source-able bash prelude and utils

dongcarl committed:
guix: Adapt guix-build to prelude, restructure hier

dongcarl committed:
guix: Add troubleshooting documentation entries

dongcarl committed:
guix: Supply --link-profile

dongcarl committed:
guix: More thoroughly control native toolchain

dongcarl committed:
guix: Add early health check for guix-daemon

dongcarl committed:
guix: Fallback to local build for substitute-enabled Guix users

dongcarl committed:
guix: Use clang-toolchain instead of clang

dongcarl committed:
guix: Remove guix-build.sh filename extension

dongcarl committed:
guix: Remove codesign_allocate+pagestuff from unsigned tarball

sipa commented:
```
$ sha256sum $(find -type f | LANG=C sort)
54f46bc056b2d38160d5fde32590a76148e15c37d67d34fc65390ba3f371ea97  ./aarch64-linux-gnu/bitcoin-61194db48c87-aarch64-linux-gnu-debug.tar.gz
9d23600991e1b40aeee8df638e8140183b310ea8fc9a4b080fe1c0a895c173f7  ./aarch64-linux-gnu/bitcoin-61194db48c87-aarch64-linux-gnu.tar.gz
f92f44566ebd3b315d4fbac110b5583d38258d90243bb0690046cebcbd6db85c  ./arm-linux-gnueabihf/bitcoin-61194db48c87-arm-linux-gnueabihf-debug.tar.gz
a9a5656b331988051732f283612ef5c5451c7fba5563af65adb4765e00189bca  ./arm-linux-gnueabihf/bitcoin-61194db48c87-arm-linux-gnueabihf.tar.gz
56d4856462f6c627910da9932d4477d2f09335a2e21a67fc02161b6e87324f43  ./dist-archive/bitcoin-61194db48c87.tar.gz
fd9573284e95c17b023ff34a9c1e63ef2774d7a9d56148dd47e619e9db3fd552  ./powerpc64-linux-gnu/bitcoin-61194db48c87-powerpc64-linux-gnu-debug.tar.gz
997cb276dee01ab9aa6c8f742a4fd4550570c9bfddfe016abde0c0a5ed96f626  ./powerpc64-linux-gnu/bitcoin-61194db48c87-powerpc64-linux-gnu.tar.gz
33037bfa1cf621fdffc7db9d79924fa20396f63dce7c6d75051a4aa8ccb83d60  ./powerpc64le-linux-gnu/bitcoin-61194db48c87-powerpc64le-linux-gnu-debug.tar.gz
c7f2d79a785162453e825345dff38cb53e4fea2f1752375b480138397556a78c  ./powerpc64le-linux-gnu/bitcoin-61194db48c87-powerpc64le-linux-gnu.tar.gz
2df5d6a009ae1ba4b87bb68190fbcc0144d779ef0658582510087ab03a3138df  ./riscv64-linux-gnu/bitcoin-61194db48c87-riscv64-linux-gnu-debug.tar.gz
fe4518267aff57ee69ed8333654661dbe3672d53149a38880ba09226b04bcd56  ./riscv64-linux-gnu/bitcoin-61194db48c87-riscv64-linux-gnu.tar.gz
cee31dfd3a107ac2a83d4cebea19550b3c3aa9d6996ecfd956072cbf3ac12061  ./x86_64-apple-darwin18/bitcoin-61194db48c87-osx-unsigned.dmg
fd1e1116adac330def0b285763262941f23f9d792dc92bb07e68620f1b6dc556  ./x86_64-apple-darwin18/bitcoin-61194db48c87-osx-unsigned.tar.gz
12194e2c4504a63329b082c26543fff112f3ebc184b2aadbbfb57779d1968047  ./x86_64-apple-darwin18/bitcoin-61194db48c87-osx64.tar.gz
81006656851795b0414e523520ec3d222d9033bc84c68354332632b8536e4d87  ./x86_64-linux-gnu/bitcoin-61194db48c87-x86_64-linux-gnu-debug.tar.gz
678dfae6336edefe90d048c05335f699bb7fe27b7420d47674bc3125e79f6b3b  ./x86_64-linux-gnu/bitcoin-61194db48c87-x86_64-linux-gnu.tar.gz
56a1eec1494674294015bbfa7bba3fa65c95eb7c4984e366483d3e0c6795b887  ./x86_64-w64-mingw32/bitcoin-61194db48c87-win-unsigned.tar.gz
01f04e86a8371fda4971914d06a775ed00a1122713427b8e591aadef675a5e31  ./x86_64-w64-mingw32/bitcoin-61194db48c87-win64-debug.zip
1f83850e8e22ef93c2731ff99b1c0b7ae2cb70a8712d3b04c7bb4a5080d91c39  ./x86_64-w64-mingw32/bitcoin-61194db48c87-win64-setup-unsigned.exe
408d22aea8d5f22ade221ed9188ea859ef1f2395b7d74d817855e090a2328321  ./x86_64-w64-mingw32/bitcoin-61194db48c87-win64.zip
```

dongcarl commented on contrib/guix/guix-build:
```diff
@@ -216,15 +227,15 @@ for host in $HOSTS; do
           ...bind-mounted in container to: '$(DISTSRC_BASE=/distsrc-base && distsrc_for_host "$HOST")'
-      ...outputting in: '${OUTDIR:?not set}'
-          ...bind-mounted in container to: '/outdir'
+      ...outdirting in: '$(outdir_for_host "$HOST")'
```
😬 Uhh... Over-eager find-and-replace strikes again...

sipa replied:
This is still not fixed, FWIW.

fanquake commented:
My outdirt. Looks like there's still a diff in the osx  unsigned tarball:
```bash
find output -type f -name *$(git rev-parse --short HEAD)*.* -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
54f46bc056b2d38160d5fde32590a76148e15c37d67d34fc65390ba3f371ea97  output/aarch64-linux-gnu/bitcoin-61194db48c87-aarch64-linux-gnu-debug.tar.gz
9d23600991e1b40aeee8df638e8140183b310ea8fc9a4b080fe1c0a895c173f7  output/aarch64-linux-gnu/bitcoin-61194db48c87-aarch64-linux-gnu.tar.gz
f92f44566ebd3b315d4fbac110b5583d38258d90243bb0690046cebcbd6db85c  output/arm-linux-gnueabihf/bitcoin-61194db48c87-arm-linux-gnueabihf-debug.tar.gz
a9a5656b331988051732f283612ef5c5451c7fba5563af65adb4765e00189bca  output/arm-linux-gnueabihf/bitcoin-61194db48c87-arm-linux-gnueabihf.tar.gz
56d4856462f6c627910da9932d4477d2f09335a2e21a67fc02161b6e87324f43  output/dist-archive/bitcoin-61194db48c87.tar.gz
fd9573284e95c17b023ff34a9c1e63ef2774d7a9d56148dd47e619e9db3fd552  output/powerpc64-linux-gnu/bitcoin-61194db48c87-powerpc64-linux-gnu-debug.tar.gz
997cb276dee01ab9aa6c8f742a4fd4550570c9bfddfe016abde0c0a5ed96f626  output/powerpc64-linux-gnu/bitcoin-61194db48c87-powerpc64-linux-gnu.tar.gz
33037bfa1cf621fdffc7db9d79924fa20396f63dce7c6d75051a4aa8ccb83d60  output/powerpc64le-linux-gnu/bitcoin-61194db48c87-powerpc64le-linux-gnu-debug.tar.gz
c7f2d79a785162453e825345dff38cb53e4fea2f1752375b480138397556a78c  output/powerpc64le-linux-gnu/bitcoin-61194db48c87-powerpc64le-linux-gnu.tar.gz
2df5d6a009ae1ba4b87bb68190fbcc0144d779ef0658582510087ab03a3138df  output/riscv64-linux-gnu/bitcoin-61194db48c87-riscv64-linux-gnu-debug.tar.gz
fe4518267aff57ee69ed8333654661dbe3672d53149a38880ba09226b04bcd56  output/riscv64-linux-gnu/bitcoin-61194db48c87-riscv64-linux-gnu.tar.gz
cee31dfd3a107ac2a83d4cebea19550b3c3aa9d6996ecfd956072cbf3ac12061  output/x86_64-apple-darwin18/bitcoin-61194db48c87-osx-unsigned.dmg
5e1237b7197c2bd9f163371c5fcdb03f111db9bcfb62f178de303add7f2c1528  output/x86_64-apple-darwin18/bitcoin-61194db48c87-osx-unsigned.tar.gz
12194e2c4504a63329b082c26543fff112f3ebc184b2aadbbfb57779d1968047  output/x86_64-apple-darwin18/bitcoin-61194db48c87-osx64.tar.gz
81006656851795b0414e523520ec3d222d9033bc84c68354332632b8536e4d87  output/x86_64-linux-gnu/bitcoin-61194db48c87-x86_64-linux-gnu-debug.tar.gz
678dfae6336edefe90d048c05335f699bb7fe27b7420d47674bc3125e79f6b3b  output/x86_64-linux-gnu/bitcoin-61194db48c87-x86_64-linux-gnu.tar.gz
56a1eec1494674294015bbfa7bba3fa65c95eb7c4984e366483d3e0c6795b887  output/x86_64-w64-mingw32/bitcoin-61194db48c87-win-unsigned.tar.gz
01f04e86a8371fda4971914d06a775ed00a1122713427b8e591aadef675a5e31  output/x86_64-w64-mingw32/bitcoin-61194db48c87-win64-debug.zip
1f83850e8e22ef93c2731ff99b1c0b7ae2cb70a8712d3b04c7bb4a5080d91c39  output/x86_64-w64-mingw32/bitcoin-61194db48c87-win64-setup-unsigned.exe
408d22aea8d5f22ade221ed9188ea859ef1f2395b7d74d817855e090a2328321  output/x86_64-w64-mingw32/bitcoin-61194db48c87-win64.zip
```

dongcarl commented:
Mine!
```
$ find guix-build-61194db48c87/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
54f46bc056b2d38160d5fde32590a76148e15c37d67d34fc65390ba3f371ea97  guix-build-61194db48c87/output/aarch64-linux-gnu/bitcoin-61194db48c87-aarch64-linux-gnu-debug.tar.gz
9d23600991e1b40aeee8df638e8140183b310ea8fc9a4b080fe1c0a895c173f7  guix-build-61194db48c87/output/aarch64-linux-gnu/bitcoin-61194db48c87-aarch64-linux-gnu.tar.gz
f92f44566ebd3b315d4fbac110b5583d38258d90243bb0690046cebcbd6db85c  guix-build-61194db48c87/output/arm-linux-gnueabihf/bitcoin-61194db48c87-arm-linux-gnueabihf-debug.tar.gz
a9a5656b331988051732f283612ef5c5451c7fba5563af65adb4765e00189bca  guix-build-61194db48c87/output/arm-linux-gnueabihf/bitcoin-61194db48c87-arm-linux-gnueabihf.tar.gz
56d4856462f6c627910da9932d4477d2f09335a2e21a67fc02161b6e87324f43  guix-build-61194db48c87/output/dist-archive/bitcoin-61194db48c87.tar.gz
fd9573284e95c17b023ff34a9c1e63ef2774d7a9d56148dd47e619e9db3fd552  guix-build-61194db48c87/output/powerpc64-linux-gnu/bitcoin-61194db48c87-powerpc64-linux-gnu-debug.tar.gz
997cb276dee01ab9aa6c8f742a4fd4550570c9bfddfe016abde0c0a5ed96f626  guix-build-61194db48c87/output/powerpc64-linux-gnu/bitcoin-61194db48c87-powerpc64-linux-gnu.tar.gz
33037bfa1cf621fdffc7db9d79924fa20396f63dce7c6d75051a4aa8ccb83d60  guix-build-61194db48c87/output/powerpc64le-linux-gnu/bitcoin-61194db48c87-powerpc64le-linux-gnu-debug.tar.gz
c7f2d79a785162453e825345dff38cb53e4fea2f1752375b480138397556a78c  guix-build-61194db48c87/output/powerpc64le-linux-gnu/bitcoin-61194db48c87-powerpc64le-linux-gnu.tar.gz
2df5d6a009ae1ba4b87bb68190fbcc0144d779ef0658582510087ab03a3138df  guix-build-61194db48c87/output/riscv64-linux-gnu/bitcoin-61194db48c87-riscv64-linux-gnu-debug.tar.gz
fe4518267aff57ee69ed8333654661dbe3672d53149a38880ba09226b04bcd56  guix-build-61194db48c87/output/riscv64-linux-gnu/bitcoin-61194db48c87-riscv64-linux-gnu.tar.gz
cee31dfd3a107ac2a83d4cebea19550b3c3aa9d6996ecfd956072cbf3ac12061  guix-build-61194db48c87/output/x86_64-apple-darwin18/bitcoin-61194db48c87-osx-unsigned.dmg
0173b1cf6992390602433aa1d99ebf3e33d47cbad591a6b11f23350cfb4b9b5e  guix-build-61194db48c87/output/x86_64-apple-darwin18/bitcoin-61194db48c87-osx-unsigned.tar.gz
12194e2c4504a63329b082c26543fff112f3ebc184b2aadbbfb57779d1968047  guix-build-61194db48c87/output/x86_64-apple-darwin18/bitcoin-61194db48c87-osx64.tar.gz
81006656851795b0414e523520ec3d222d9033bc84c68354332632b8536e4d87  guix-build-61194db48c87/output/x86_64-linux-gnu/bitcoin-61194db48c87-x86_64-linux-gnu-debug.tar.gz
678dfae6336edefe90d048c05335f699bb7fe27b7420d47674bc3125e79f6b3b  guix-build-61194db48c87/output/x86_64-linux-gnu/bitcoin-61194db48c87-x86_64-linux-gnu.tar.gz
56a1eec1494674294015bbfa7bba3fa65c95eb7c4984e366483d3e0c6795b887  guix-build-61194db48c87/output/x86_64-w64-mingw32/bitcoin-61194db48c87-win-unsigned.tar.gz
01f04e86a8371fda4971914d06a775ed00a1122713427b8e591aadef675a5e31  guix-build-61194db48c87/output/x86_64-w64-mingw32/bitcoin-61194db48c87-win64-debug.zip
1f83850e8e22ef93c2731ff99b1c0b7ae2cb70a8712d3b04c7bb4a5080d91c39  guix-build-61194db48c87/output/x86_64-w64-mingw32/bitcoin-61194db48c87-win64-setup-unsigned.exe
408d22aea8d5f22ade221ed9188ea859ef1f2395b7d74d817855e090a2328321  guix-build-61194db48c87/output/x86_64-w64-mingw32/bitcoin-61194db48c87-win64.zip
```
I'm going to guess `dmg` is causing it... If one of you can upload I'm going to get to the bottom of this.

fanquake commented:
> If one of you can upload I'm going to get to the bottom of this.

Uploaded here: https://github.com/fanquake/build_artifacts/releases/tag/guix_21375

dongcarl committed:
depends: libdmg-hfsplus: Skip CMake RPATH patching

CMake's RPATH patching apparently causes non-reproducibility in the
executables which are produced, manifesting in a difference in padding
in the .dynstr section (we found this while investigating
non-reproducibility in the "dmg" tool). This RPATH patching can be
safely skipped for executables which don't depend on internal shared
libraries.

Documentation sources:
1. https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/RPATH-handling
2. https://reproducible-builds.org/docs/deterministic-build-systems/#cmake-notes

Prior debugging art:
1. https://stackoverflow.com/questions/63438206/cmake-g-reproducible-build-issue-with-changing-build-path
2. https://github.com/NXPmicro/mfgtools/pull/229/files

dongcarl commented:
Testing a hypothesis: @sipa @fanquake could one of you build 8170456b5b0b5f20e869e9a59b694383c57de28e with HOSTS='x86_64-apple-darwin18'?

sipa commented:
@dongcarl I got:

```
1bbdaaeda79420b5d4b33de5490344e9526783ca43dbf31e11c02b34a4761685  ./dist-archive/bitcoin-8170456b5b0b.tar.gz
417cd25618d10ec7efe6575ac79cf5265dfe689219113cd19a834b61deeb275e  ./x86_64-apple-darwin18/bitcoin-8170456b5b0b-osx-unsigned.dmg
f549b232b3eb6d45fcd4f8939b2ad3e80033c470d907a119ea8ec165dafa43f2  ./x86_64-apple-darwin18/bitcoin-8170456b5b0b-osx-unsigned.tar.gz
21a976a7647679f34a6afc54a25d0cd85d9a44d33891adab01153c992f8da512  ./x86_64-apple-darwin18/bitcoin-8170456b5b0b-osx64.tar.gz
```

dongcarl commented:
Great! Looks like we have reproducibility. I'm experimenting a bit with building `dmg` as a static executable right now, which I think is the right move. Any binary we put in our output should "err on the side on portability," and in this case, since `dmg` doesn't need NSS or any networking, we can just make it fully static.

hebasto commented:
```
$ find guix-build-8170456b5b0b/output -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
1bbdaaeda79420b5d4b33de5490344e9526783ca43dbf31e11c02b34a4761685  guix-build-8170456b5b0b/output/dist-archive/bitcoin-8170456b5b0b.tar.gz
417cd25618d10ec7efe6575ac79cf5265dfe689219113cd19a834b61deeb275e  guix-build-8170456b5b0b/output/x86_64-apple-darwin18/bitcoin-8170456b5b0b-osx-unsigned.dmg
f549b232b3eb6d45fcd4f8939b2ad3e80033c470d907a119ea8ec165dafa43f2  guix-build-8170456b5b0b/output/x86_64-apple-darwin18/bitcoin-8170456b5b0b-osx-unsigned.tar.gz
21a976a7647679f34a6afc54a25d0cd85d9a44d33891adab01153c992f8da512  guix-build-8170456b5b0b/output/x86_64-apple-darwin18/bitcoin-8170456b5b0b-osx64.tar.gz
```

laanwj commented:
What is the rationale to remove the `.sh` from the build script filename?

It seems to me that the convention in the rest of the repository is to keep the .py/.sh extension for scripts, even the main units.

dongcarl commented:
> What is the rationale for removing the `.sh` from the build script filename?

I was thinking that we might change the scripting language for these scripts in the future, or perhaps just have a python wrapper which interprets arguments. I that case, I wouldn't want to have everyone change their invocation again, therefore I omitted the `.sh`. I'm happy to do either way though!

laanwj commented:
My output is:
```
$ cd guix-build-*/output/ && find * -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
fa28462f5c23413fd8c984f46a36238e69cd3e33871cedeae70c54f9a5977363  aarch64-linux-gnu/bitcoin-8170456b5b0b-aarch64-linux-gnu-debug.tar.gz
14d40d7faa2775a7cf91a90d1da83d0928e75d44728f1e2b7e7164708a6be9c8  aarch64-linux-gnu/bitcoin-8170456b5b0b-aarch64-linux-gnu.tar.gz
1c98c9fb41af4afb180122b8b9af57148344d25c029e20c5ccd2ec8e30b44028  arm-linux-gnueabihf/bitcoin-8170456b5b0b-arm-linux-gnueabihf-debug.tar.gz
0237c2e46ef75f69fbed6050d8ee7795be754249767d1c773fe3fc34cdabfbba  arm-linux-gnueabihf/bitcoin-8170456b5b0b-arm-linux-gnueabihf.tar.gz
1bbdaaeda79420b5d4b33de5490344e9526783ca43dbf31e11c02b34a4761685  dist-archive/bitcoin-8170456b5b0b.tar.gz
6fe902c22accad9476a98e41d9421cde20618285a1052b7935053900ad98d774  powerpc64-linux-gnu/bitcoin-8170456b5b0b-powerpc64-linux-gnu-debug.tar.gz
3d242604ab685ef6053ff63b1a87b6aa6e8e4a45611244b0046f0621f9bb0cb2  powerpc64-linux-gnu/bitcoin-8170456b5b0b-powerpc64-linux-gnu.tar.gz
22d37162e34981fb7af66986f5067a7af15f53616fdbbceefb0e6465ae53b037  powerpc64le-linux-gnu/bitcoin-8170456b5b0b-powerpc64le-linux-gnu-debug.tar.gz
946c6b53f2874408ba43f67d4b8debafb8fc7767f11a98ce66e0bfd3fa2b10bf  powerpc64le-linux-gnu/bitcoin-8170456b5b0b-powerpc64le-linux-gnu.tar.gz
981e41c0a22b880f1e375c268b1a2876af3f3f2662205b3d2af97b8809e6bc8a  riscv64-linux-gnu/bitcoin-8170456b5b0b-riscv64-linux-gnu-debug.tar.gz
84ed7e0671ab626f015e13a0650ad4ea22a7d60e9e958611cfe26527e627a6b8  riscv64-linux-gnu/bitcoin-8170456b5b0b-riscv64-linux-gnu.tar.gz
417cd25618d10ec7efe6575ac79cf5265dfe689219113cd19a834b61deeb275e  x86_64-apple-darwin18/bitcoin-8170456b5b0b-osx-unsigned.dmg
f549b232b3eb6d45fcd4f8939b2ad3e80033c470d907a119ea8ec165dafa43f2  x86_64-apple-darwin18/bitcoin-8170456b5b0b-osx-unsigned.tar.gz
21a976a7647679f34a6afc54a25d0cd85d9a44d33891adab01153c992f8da512  x86_64-apple-darwin18/bitcoin-8170456b5b0b-osx64.tar.gz
929b616cf6acc9f1e327fc0af9c617573b167a25057fb4e2dc69e58027b41129  x86_64-linux-gnu/bitcoin-8170456b5b0b-x86_64-linux-gnu-debug.tar.gz
607c2cfea2a2c9d471d9de8bdea2293dc3058cf23098777086e18f27fbc7ce5f  x86_64-linux-gnu/bitcoin-8170456b5b0b-x86_64-linux-gnu.tar.gz
cd06f72d2a0d39f215f90070893e1ac638b8673281e296d7d2f2f29757e323d6  x86_64-w64-mingw32/bitcoin-8170456b5b0b-win-unsigned.tar.gz
e54a7adc915dcb9873c6a7b46f657061c3dc2c270f2ba9e0e218b2cbddd09cc5  x86_64-w64-mingw32/bitcoin-8170456b5b0b-win64-debug.zip
131816b9b9603a4a3e1ff035015f42d6839d1d14528a0d94f6e2052a00323933  x86_64-w64-mingw32/bitcoin-8170456b5b0b-win64-setup-unsigned.exe
94c51abeedceeb9c777fba4d18156055494ea350d862b994b7026acb0c96dc16  x86_64-w64-mingw32/bitcoin-8170456b5b0b-win64.zip
```

I checked that the apple ones match.

> I was thinking that we might change the scripting language for these scripts in the future, or perhaps just have a python wrapper which interprets arguments.

Okay, yes, implementing it in Python sounds great to me tbh. I do not like complex shell scripts. And yes, having to update all workflow documentation for a new extension would be inconvenient. Fair enough I'm fine with keeping it like this it just surprised me coming out of the blue.

dongcarl committed:
guix: Build dmg as a static binary

This relatively easy change eliminates all runtime dependencies (except
for the kernel) for dmg, which is the only native build tool that gets
put in our output tarballs.

This allows much more flexibility when constructing the codesigning
environment, and is much more robust.

dongcarl commented:
Pushed 8170456b5b0b5f20e869e9a59b694383c57de28e -> e7e778bd13e6c696369ffdf317054f10679e8ac2
- Supply `--fallback` to `guix time-machine` as well
- Add notes w/re CMake RPATH patching in commit message
- Add commit to build `dmg` as a static binary

sipa commented:
```
$ sha256sum $(find -type f | LANG=C sort)
27cfd548cffca972b1efb4e9e8063eb20935420e0d24cac1ca4b1816d5c14e15  ./aarch64-linux-gnu/bitcoin-e7e778bd13e6-aarch64-linux-gnu-debug.tar.gz
806c1532a9329507442bc07d142ca520523cde9caa813587620d77a6427c3ec1  ./aarch64-linux-gnu/bitcoin-e7e778bd13e6-aarch64-linux-gnu.tar.gz
ede546c8bf428c59b201b95b0063c0740bd1926ef7d5b0c51481250dc616bd9e  ./arm-linux-gnueabihf/bitcoin-e7e778bd13e6-arm-linux-gnueabihf-debug.tar.gz
e34701be5a5c06284474c20c2dfa69044cf3f2d4f83075d6e0d3cc3499e62079  ./arm-linux-gnueabihf/bitcoin-e7e778bd13e6-arm-linux-gnueabihf.tar.gz
535636c8139bb605f318abd81a82734f6de1499804d18f2bfc94ae83a24dca83  ./dist-archive/bitcoin-e7e778bd13e6.tar.gz
ea96a5e4ceb753cc7730756ef9f3614f99c52a85ca3dcd3c27be3a839e5786aa  ./powerpc64-linux-gnu/bitcoin-e7e778bd13e6-powerpc64-linux-gnu-debug.tar.gz
d5c67cb9ae7141cfddb84dea72e2dbdd75c6973c5a280e2592c8a030f754a7d3  ./powerpc64-linux-gnu/bitcoin-e7e778bd13e6-powerpc64-linux-gnu.tar.gz
11e5850addf5e36ea1a88bdbfb95eaf0849dffad2a02ea9a9f6c6ed298697906  ./powerpc64le-linux-gnu/bitcoin-e7e778bd13e6-powerpc64le-linux-gnu-debug.tar.gz
f2c9b9a3bf1ca9e6534f1e6b257726fe4f54887f983ed106f803e1c20b391969  ./powerpc64le-linux-gnu/bitcoin-e7e778bd13e6-powerpc64le-linux-gnu.tar.gz
e956d36f43dc165f3d891ec15f444c2b9118db62731c44d7fb7e02ce7adb5627  ./riscv64-linux-gnu/bitcoin-e7e778bd13e6-riscv64-linux-gnu-debug.tar.gz
a1d463ad0d0d30300040359e08854e83802484f308b7d05cf8db91488369e70b  ./riscv64-linux-gnu/bitcoin-e7e778bd13e6-riscv64-linux-gnu.tar.gz
768b4d350479b41023a399b039b4b41740e0665ca74f932202fd6b7778aff2fc  ./x86_64-apple-darwin18/bitcoin-e7e778bd13e6-osx-unsigned.dmg
5445a0ec5fd2499037743c19fc5333e017ca5a580c7577b7dc3c0bd797ca150e  ./x86_64-apple-darwin18/bitcoin-e7e778bd13e6-osx-unsigned.tar.gz
19c44b80bca204dedc9930d8185694462c32ded12858c0cc60d2d827a65f164c  ./x86_64-apple-darwin18/bitcoin-e7e778bd13e6-osx64.tar.gz
d73c928e4526609d965af1deeebd024446cf1c805c5e1c6c3b4b6d3f83b544bd  ./x86_64-linux-gnu/bitcoin-e7e778bd13e6-x86_64-linux-gnu-debug.tar.gz
6451444951c00b4069470fe59c56381605a00c81bdf4704ee7e864fa1874c351  ./x86_64-linux-gnu/bitcoin-e7e778bd13e6-x86_64-linux-gnu.tar.gz
4f74734e93f3c24d77bbe05086b19c2196b6838cebb216379fd490a58eea096c  ./x86_64-w64-mingw32/bitcoin-e7e778bd13e6-win-unsigned.tar.gz
cbe39967d66df4234ac399326e918d08fbb72ce3df2c172b66074de69c747b5c  ./x86_64-w64-mingw32/bitcoin-e7e778bd13e6-win64-debug.zip
6b2dfd59e8be58fb14ae92be27d47587681e890d816f6063dc3177ca21ee9ee5  ./x86_64-w64-mingw32/bitcoin-e7e778bd13e6-win64-setup-unsigned.exe
e6ac8c8044ab6ace19976000922d585f56681918e08095712f26775e58904bdb  ./x86_64-w64-mingw32/bitcoin-e7e778bd13e6-win64.zip
```

hebasto commented on contrib/guix/guix-build:
```diff
@@ -2,22 +2,26 @@
-    fi
-done
+
+check_tools cat mkdir make git guix
```
`cat` and `git` already checked in the sourced `libexec/prelude.bash`.
Maybe combine both `check_tools` into one?

dongcarl replied:
The check is quite cheap, and only attest to the tools being used in the current script. Instead of trying to figure out how to optimize it, I think we can just leave it be if it's redundant. 😄

hebasto commented on contrib/guix/README.md:
```diff
@@ -265,6 +265,60 @@ To use dongcarl's substitute server for Bitcoin Core builds after having
+times, it may be `/tmp/...drv-1` or `/tmp/...drv-2`. Always consult the build
+failure output for the most accurate, up-to-date information.
+
+#### python(-minimal): [Errno 84] Invalid or incomplete multibyte or wide character
```
Maybe inline with the previous "Derivation failed to build" item
```suggestion
### python(-minimal): [Errno 84] Invalid or incomplete multibyte or wide character
```
?

dongcarl replied:
This is a common "Derivation failed to build" error, which is why I made it underneath it hierarchically

hebasto commented:
I'd suggest to split this PR in two separated parts:
1) bash scripts optimizations and fixes, dir structure change, docs
2) dependencies build stuff, patches etc

It would make each part more focused, and easier to review and test, no?

dongcarl commented:
> It would make each part more focused, and easier to review and test, no?
That may be true, but I think a lot of people have already tested the whole bundle here, so I want to keep it this way if possible :-)

If there are parts that you find especially hard to parse/review, I'd be happy to explain more!

dongcarl commented:
Pushed e7e778bd13e6c696369ffdf317054f10679e8ac2 -> edf5c8a8182e30862c7d62f11147ead31e029f04
- Rebased over master

```sh
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
4f75b7e4f4037d47b405025a23f7ec762a6b16645858c9f547f95ea22651b9fa  guix-build-edf5c8a8182e/output/aarch64-linux-gnu/bitcoin-edf5c8a8182e-aarch64-linux-gnu-debug.tar.gz
d56a231299ee283faa69044264a40673095817bf6591e260baa317684b093af1  guix-build-edf5c8a8182e/output/aarch64-linux-gnu/bitcoin-edf5c8a8182e-aarch64-linux-gnu.tar.gz
0d740d6956c6cbc9b3c003665811780cad26f7264909f7666932f11b3965e6fb  guix-build-edf5c8a8182e/output/arm-linux-gnueabihf/bitcoin-edf5c8a8182e-arm-linux-gnueabihf-debug.tar.gz
a507d8ec33c10fc8d8475882da50c344e6a9c8e0049371634f59368d7c7a3be4  guix-build-edf5c8a8182e/output/arm-linux-gnueabihf/bitcoin-edf5c8a8182e-arm-linux-gnueabihf.tar.gz
046de55652c7825aecd059c9dfaede839d48049e836e0710e8550e435cd20aa4  guix-build-edf5c8a8182e/output/dist-archive/bitcoin-edf5c8a8182e.tar.gz
039feec1265b39a39e7f827d7df16c8096c3f01a330bbf603de474bfe639a2a1  guix-build-edf5c8a8182e/output/powerpc64-linux-gnu/bitcoin-edf5c8a8182e-powerpc64-linux-gnu-debug.tar.gz
b0073d2f5cd47fbc50852261bb214f0cc360ea02e564d2775fe38a93358db176  guix-build-edf5c8a8182e/output/powerpc64-linux-gnu/bitcoin-edf5c8a8182e-powerpc64-linux-gnu.tar.gz
cef44d9ef73e6469f03cb24210249559eaa6e36cd4b558a64df92c9318ee12f8  guix-build-edf5c8a8182e/output/powerpc64le-linux-gnu/bitcoin-edf5c8a8182e-powerpc64le-linux-gnu-debug.tar.gz
0805455b000b798ff75110c1a96afd5d898fe8c0aa866f5fe49c04d153f49381  guix-build-edf5c8a8182e/output/powerpc64le-linux-gnu/bitcoin-edf5c8a8182e-powerpc64le-linux-gnu.tar.gz
40268bbdca2009cfe0bce3bf062764447958c6904342183183249817131fe739  guix-build-edf5c8a8182e/output/riscv64-linux-gnu/bitcoin-edf5c8a8182e-riscv64-linux-gnu-debug.tar.gz
f40195d1d041e87cbdf85bbb8feb496e4ec8182b6f892ef883636a8c96b4c36d  guix-build-edf5c8a8182e/output/riscv64-linux-gnu/bitcoin-edf5c8a8182e-riscv64-linux-gnu.tar.gz
2bc6df7878ca8afe8798bdf61057e5932ebe250eb010f34fb09d5630e0102e8c  guix-build-edf5c8a8182e/output/x86_64-apple-darwin18/bitcoin-edf5c8a8182e-osx-unsigned.dmg
035334329a682859e79613c13acd85981ed767f0cf65a896bfabce8ae18f71b4  guix-build-edf5c8a8182e/output/x86_64-apple-darwin18/bitcoin-edf5c8a8182e-osx-unsigned.tar.gz
689d0c40663b895e272442722ce86416347e1b1d540e948799cb765ae1281988  guix-build-edf5c8a8182e/output/x86_64-apple-darwin18/bitcoin-edf5c8a8182e-osx64.tar.gz
5addcb7e2a0ec00be93ca966467d5e9fe3b74ed4ccf5bbb070bac1d3ab1d63e6  guix-build-edf5c8a8182e/output/x86_64-linux-gnu/bitcoin-edf5c8a8182e-x86_64-linux-gnu-debug.tar.gz
638bd4559bf2a565279b1fd480a35897f552edcf23f7389cb4f694b03ee5e894  guix-build-edf5c8a8182e/output/x86_64-linux-gnu/bitcoin-edf5c8a8182e-x86_64-linux-gnu.tar.gz
e5b1cef87f107c5526a2ef6924b08e586129696774bb812aee653c3fad5be8a3  guix-build-edf5c8a8182e/output/x86_64-w64-mingw32/bitcoin-edf5c8a8182e-win-unsigned.tar.gz
707a0762678fcbbb3199a645cbf5dc47b787558a18c4e24695eeef8c6563a6bf  guix-build-edf5c8a8182e/output/x86_64-w64-mingw32/bitcoin-edf5c8a8182e-win64-debug.zip
6cdc35406bf1b2fbd4fc59103a668486b7b88f4677abde66728b43d709071c52  guix-build-edf5c8a8182e/output/x86_64-w64-mingw32/bitcoin-edf5c8a8182e-win64-setup-unsigned.exe
fdd08d5fde953a380b1e9a8d714ab4868a0172f809a84ae9f23cc5d4ac19f07b  guix-build-edf5c8a8182e/output/x86_64-w64-mingw32/bitcoin-edf5c8a8182e-win64.zip
```

sipa commented:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
4f75b7e4f4037d47b405025a23f7ec762a6b16645858c9f547f95ea22651b9fa  guix-build-edf5c8a8182e/output/aarch64-linux-gnu/bitcoin-edf5c8a8182e-aarch64-linux-gnu-debug.tar.gz
d56a231299ee283faa69044264a40673095817bf6591e260baa317684b093af1  guix-build-edf5c8a8182e/output/aarch64-linux-gnu/bitcoin-edf5c8a8182e-aarch64-linux-gnu.tar.gz
0d740d6956c6cbc9b3c003665811780cad26f7264909f7666932f11b3965e6fb  guix-build-edf5c8a8182e/output/arm-linux-gnueabihf/bitcoin-edf5c8a8182e-arm-linux-gnueabihf-debug.tar.gz
a507d8ec33c10fc8d8475882da50c344e6a9c8e0049371634f59368d7c7a3be4  guix-build-edf5c8a8182e/output/arm-linux-gnueabihf/bitcoin-edf5c8a8182e-arm-linux-gnueabihf.tar.gz
046de55652c7825aecd059c9dfaede839d48049e836e0710e8550e435cd20aa4  guix-build-edf5c8a8182e/output/dist-archive/bitcoin-edf5c8a8182e.tar.gz
039feec1265b39a39e7f827d7df16c8096c3f01a330bbf603de474bfe639a2a1  guix-build-edf5c8a8182e/output/powerpc64-linux-gnu/bitcoin-edf5c8a8182e-powerpc64-linux-gnu-debug.tar.gz
b0073d2f5cd47fbc50852261bb214f0cc360ea02e564d2775fe38a93358db176  guix-build-edf5c8a8182e/output/powerpc64-linux-gnu/bitcoin-edf5c8a8182e-powerpc64-linux-gnu.tar.gz
cef44d9ef73e6469f03cb24210249559eaa6e36cd4b558a64df92c9318ee12f8  guix-build-edf5c8a8182e/output/powerpc64le-linux-gnu/bitcoin-edf5c8a8182e-powerpc64le-linux-gnu-debug.tar.gz
0805455b000b798ff75110c1a96afd5d898fe8c0aa866f5fe49c04d153f49381  guix-build-edf5c8a8182e/output/powerpc64le-linux-gnu/bitcoin-edf5c8a8182e-powerpc64le-linux-gnu.tar.gz
40268bbdca2009cfe0bce3bf062764447958c6904342183183249817131fe739  guix-build-edf5c8a8182e/output/riscv64-linux-gnu/bitcoin-edf5c8a8182e-riscv64-linux-gnu-debug.tar.gz
f40195d1d041e87cbdf85bbb8feb496e4ec8182b6f892ef883636a8c96b4c36d  guix-build-edf5c8a8182e/output/riscv64-linux-gnu/bitcoin-edf5c8a8182e-riscv64-linux-gnu.tar.gz
2bc6df7878ca8afe8798bdf61057e5932ebe250eb010f34fb09d5630e0102e8c  guix-build-edf5c8a8182e/output/x86_64-apple-darwin18/bitcoin-edf5c8a8182e-osx-unsigned.dmg
035334329a682859e79613c13acd85981ed767f0cf65a896bfabce8ae18f71b4  guix-build-edf5c8a8182e/output/x86_64-apple-darwin18/bitcoin-edf5c8a8182e-osx-unsigned.tar.gz
689d0c40663b895e272442722ce86416347e1b1d540e948799cb765ae1281988  guix-build-edf5c8a8182e/output/x86_64-apple-darwin18/bitcoin-edf5c8a8182e-osx64.tar.gz
5addcb7e2a0ec00be93ca966467d5e9fe3b74ed4ccf5bbb070bac1d3ab1d63e6  guix-build-edf5c8a8182e/output/x86_64-linux-gnu/bitcoin-edf5c8a8182e-x86_64-linux-gnu-debug.tar.gz
638bd4559bf2a565279b1fd480a35897f552edcf23f7389cb4f694b03ee5e894  guix-build-edf5c8a8182e/output/x86_64-linux-gnu/bitcoin-edf5c8a8182e-x86_64-linux-gnu.tar.gz
e5b1cef87f107c5526a2ef6924b08e586129696774bb812aee653c3fad5be8a3  guix-build-edf5c8a8182e/output/x86_64-w64-mingw32/bitcoin-edf5c8a8182e-win-unsigned.tar.gz
707a0762678fcbbb3199a645cbf5dc47b787558a18c4e24695eeef8c6563a6bf  guix-build-edf5c8a8182e/output/x86_64-w64-mingw32/bitcoin-edf5c8a8182e-win64-debug.zip
6cdc35406bf1b2fbd4fc59103a668486b7b88f4677abde66728b43d709071c52  guix-build-edf5c8a8182e/output/x86_64-w64-mingw32/bitcoin-edf5c8a8182e-win64-setup-unsigned.exe
fdd08d5fde953a380b1e9a8d714ab4868a0172f809a84ae9f23cc5d4ac19f07b  guix-build-edf5c8a8182e/output/x86_64-w64-mingw32/bitcoin-edf5c8a8182e-win64.zip
``

sipa commented on contrib/guix/README.md:
```diff
@@ -265,6 +265,60 @@ To use dongcarl's substitute server for Bitcoin Core builds after having
+is ZFS with the utf8only=on option set.
+
+It is probably best to siwtch to using tmpfs for your `$TMPDIR`. It is the
+default for most Linux distribution.
```
In commit "guix: Add troubleshooting documentation entries"

Nit: distribution**s**.

Also, and equally a nit... do you have any evidence for that statement? I don't think it's the default on any of the distributions I've used in the past (Debian, Gentoo, Ubuntu).

laanwj replied:
From what I remember, using `tmpfs` for `/tmp` is uncommon for non-embedded / non-minimal distros, because some software like *cough* bitcoin tests *cough* like to drop large amount of files there filling up memory.

dongcarl replied:
My bad! Fixed in latest push :-)

sipa commented on contrib/guix/libexec/prelude.bash:
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env bash
```
In commit "guix: Add souce-able bash prelude and utils"

Nit: commit title contains a typo

dongcarl replied:
Fixed, thanks!

sipa reviewed (commented):
ACK edf5c8a8182e30862c7d62f11147ead31e029f04. Reviewed the code changes, and performed several builds.

dongcarl commented:
Pushed edf5c8a8182e30862c7d62f11147ead31e029f04 -> 7476b46f1893a4858616d2a8456a7c43238851ed
- Addressed https://github.com/bitcoin/bitcoin/pull/21375#discussion_r606498865, https://github.com/bitcoin/bitcoin/pull/21375#discussion_r606502338

sipa commented:
ACK 7476b46f1893a4858616d2a8456a7c43238851ed

sipa commented:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
95c20e95a7fac6e0a1d2fe8ab39f882a7b92dce457bcf617beb7885608ce6d4b  guix-build-7476b46f1893/output/aarch64-linux-gnu/bitcoin-7476b46f1893-aarch64-linux-gnu-debug.tar.gz
d1da31d62a5eb8b7bd5757994b3fed344774554a5db420a9e0126717abc1a0e0  guix-build-7476b46f1893/output/aarch64-linux-gnu/bitcoin-7476b46f1893-aarch64-linux-gnu.tar.gz
901353750ddb4edf242b6f0be8f2d792f5cd7396c92e08425af0425d48edc26f  guix-build-7476b46f1893/output/arm-linux-gnueabihf/bitcoin-7476b46f1893-arm-linux-gnueabihf-debug.tar.gz
43b82d0289bfe393d4473552daa61e9969bf2bf673b909a43e33f041251b1873  guix-build-7476b46f1893/output/arm-linux-gnueabihf/bitcoin-7476b46f1893-arm-linux-gnueabihf.tar.gz
136b6db7095ad7d9deace423042b4c8555778502d0aeaf20f7824a865cdebe1d  guix-build-7476b46f1893/output/dist-archive/bitcoin-7476b46f1893.tar.gz
63c168a78021fe9564ba64a9dacadedf3fa74ce3af94b63389b581e6d2b92a91  guix-build-7476b46f1893/output/powerpc64-linux-gnu/bitcoin-7476b46f1893-powerpc64-linux-gnu-debug.tar.gz
93bed490a7d6029fd1f96edcbe0c1a6482f90cd12ce5f520f6d57a9d879d2810  guix-build-7476b46f1893/output/powerpc64-linux-gnu/bitcoin-7476b46f1893-powerpc64-linux-gnu.tar.gz
2a085f1c9ac270502bedfda1de12ca075fe55805275aae17d92872821d5ef6dd  guix-build-7476b46f1893/output/powerpc64le-linux-gnu/bitcoin-7476b46f1893-powerpc64le-linux-gnu-debug.tar.gz
73c30078d872733cfd908cdc5b3430e01b6001ff1f8454f08d1a904fd196e502  guix-build-7476b46f1893/output/powerpc64le-linux-gnu/bitcoin-7476b46f1893-powerpc64le-linux-gnu.tar.gz
9728024ae3cff4d067410c61a1024a3dc7526a4842a51a06ff2d3a1a67294833  guix-build-7476b46f1893/output/riscv64-linux-gnu/bitcoin-7476b46f1893-riscv64-linux-gnu-debug.tar.gz
f67986dc977ee4b791a505bc14e032f625713a69d9090d9d3f5d45154785dd5d  guix-build-7476b46f1893/output/riscv64-linux-gnu/bitcoin-7476b46f1893-riscv64-linux-gnu.tar.gz
32466d33c282fbb5da9d49fe46a71403b280fbfe9fd0fe33169fbec118c1b582  guix-build-7476b46f1893/output/x86_64-apple-darwin18/bitcoin-7476b46f1893-osx-unsigned.dmg
8e3cd482b143ce87075b01881f53c233e7867a0ff97ee810529d56fa6080ee72  guix-build-7476b46f1893/output/x86_64-apple-darwin18/bitcoin-7476b46f1893-osx-unsigned.tar.gz
ef7aaebeb0cc5d14aacce1fb2bf5e2ec9ab9652142ebcc5e0378c967583c778d  guix-build-7476b46f1893/output/x86_64-apple-darwin18/bitcoin-7476b46f1893-osx64.tar.gz
9559ad27c0e76373e5e9b6a068ce0e9a51031fb121c12cdab0bb0a183624efcd  guix-build-7476b46f1893/output/x86_64-linux-gnu/bitcoin-7476b46f1893-x86_64-linux-gnu-debug.tar.gz
23fba5cfd134fd35c78e05bab8207a0909acce15ffe5fb8fed7703fabe389e96  guix-build-7476b46f1893/output/x86_64-linux-gnu/bitcoin-7476b46f1893-x86_64-linux-gnu.tar.gz
73bd95215fcdb6428ad16e785502b6438f4ec02f9c3e237f7bfab213e2b79077  guix-build-7476b46f1893/output/x86_64-w64-mingw32/bitcoin-7476b46f1893-win-unsigned.tar.gz
c2b919c0a6bdf6b31c86bc52a22a69c78df2f1adfda5e0629e24c2c5accd1715  guix-build-7476b46f1893/output/x86_64-w64-mingw32/bitcoin-7476b46f1893-win64-debug.zip
8add7726bc390f177ea67ba7fb5b6697e9f51cf1c7c82e87ef89163fcfd62834  guix-build-7476b46f1893/output/x86_64-w64-mingw32/bitcoin-7476b46f1893-win64-setup-unsigned.exe
453af78c249b5fb8341bb48850a4a886d26595f35cda7f72fea97b0f14760dc2  guix-build-7476b46f1893/output/x86_64-w64-mingw32/bitcoin-7476b46f1893-win64.zip
```

laanwj commented:
ACK 7476b46f1893a4858616d2a8456a7c43238851ed
My output (built in a GUIX VM) matches @sipa's 100%.

hebasto commented:
> Finally, we restructure the guix building hierarchy such that it looks something like:

Also see #21643.
